### PR TITLE
Nwb write fixes

### PR DESCRIPTION
**Pull request by luiztauffer**
Changes from #627 but coming from CatalystNeuro fork, so @CodyCBakerPhD and @Saksham20 can contribute

The main fixes are:

**L409**
The argument for an ElectrodeGroup is `device`, not `device_name`. Metadata with device_name would fail schema validation. This same change is reflected in other changes

**L427**
`ElectrodeGroup` is an array in metadata and, if not passed by the user, the default operation should get it from the channel groups in the recording object. Some other changes reflect this change in `defaults` to a list

**L448**
`Device` is an array in metadata

### CODE REVIEW

**Comment by luiztauffer** [comment]
@CodyCBakerPhD  tests are passing: https://github.com/catalystneuro/spikeextractors/runs/2555761593
would you like to work on other tests before merging this?

**Comment by CodyCBakerPhD** [comment]
@luiztauffer Yeah, ideally one test per issue fixed or any hidden functionalities verified

File: spikeextractors/extractors/nwbextractors/nwbextractors.py [hunk 1]
```diff
@@ -364,16 +364,21 @@ def add_devices(recording: se.RecordingExtractor, nwbfile=None, metadata: dict =
...
+
+        if metadata is None:
+            metadata = dict() 
+
+        if 'Ecephys' not in metadata:
+            metadata['Ecephys'] = dict()
+
+        if 'Device' not in metadata['Ecephys']:
+            metadata['Ecephys']['Device'] = [defaults]
+
```
Comment (CodyCBakerPhD) [inline spikeextractors/extractors/nwbextractors/nwbextractors.py:381]: While I don't like having to break a single if statement into 3 parts, I can see the issue here (though it's unmentioned in the PR post). If a user passes metadata that does not have a device field, this completely overwrites the entire dictionary as the metadata is working as a pointer at that level.

What we would really want to do here is a deep dictionary update - given that we'll soon be moving a lot of this over to nwb-conversion-tools, I'll OK this temporary change for now and we can refactor more cleanly once we gain access to all the nwb-conversion-tools utils.

File: spikeextractors/extractors/nwbextractors/nwbextractors.py [hunk 2]
```diff
@@ -411,38 +416,46 @@ def add_electrode_groups(recording: se.RecordingExtractor, nwbfile=None, metadat
...
+                            )]
                         )
                     )
                     se.NwbRecordingExtractor.add_devices(recording, nwbfile, metadata=new_device)
                     warnings.warn(f"Device \'{device_name}\' not detected in "
                                   "attempted link to electrode group! Automatically generating.")
-                electrode_group_kwargs = dict(defaults, **grp)
-                electrode_group_kwargs.pop('device_name')
+                electrode_group_kwargs = dict(defaults[0], **grp)
+                # electrode_group_kwargs.pop('device')
```
Comment (CodyCBakerPhD) [inline spikeextractors/extractors/nwbextractors/nwbextractors.py:458]:
```suggestion
```

File: spikeextractors/extractors/nwbextractors/nwbextractors.py [hunk 3]
```diff
@@ -411,38 +416,46 @@ def add_electrode_groups(recording: se.RecordingExtractor, nwbfile=None, metadat
...
+
+        if 'Ecephys' not in metadata:
+            metadata['Ecephys'] = dict()
+
+        defaults = [
+            dict(
+                name=str(group_id),
+                description="no description",
+                location="unknown",
+                device=[i.name for i in nwbfile.devices.values()][0]
```
Comment (CodyCBakerPhD) [inline spikeextractors/extractors/nwbextractors/nwbextractors.py:432]: I'm OK with this but you should note that (as apparent in the new tests) it is a rare behavior that has to be forced via the exclusion of an entire `ElectrodeGroup` key in the metadata, which is inconsistent with most defaulting use via `get_nwb_metadata`.

**Comment by bendichter** [comment]
@luiztauffer, @CodyCBakerPhD what's the status of this? How can we push this PR forward?

**Comment by bendichter** [comment]
OK, let's move this over to nwb conversion tools. Cody, I really appreciate your efforts in building tests, but we need to get this PR through.

**Comment by CodyCBakerPhD** [comment]
Now absorbed into [nwb-conversion-tools #178](https://github.com/catalystneuro/nwb-conversion-tools/pull/178).

**Comment by CodyCBakerPhD** [comment]
@luiztauffer You can close the PR but keep the branch so I can refer to the testing stuff when I port that over

**Pull request closed without merging**